apps/site/lib/site_web/controllers/stop_controller.ex

```diff
@@ -192,19 +192,44 @@ defmodule SiteWeb.StopController do
     json(conn, route_patterns_by_route_and_headsign(stop_id))
   end
 
-  @type by_route_and_headsign :: %{Route.id_t() => %{String.t() => [RoutePattern.t()]}}
+  @type headsign_info :: %{
+          required(:direction_id) => 0 | 1,
+          required(:route_patterns) => [RoutePattern.t()]
+        }
+  @type by_route_and_headsign :: %{Route.id_t() => %{String.t() => headsign_info}}
   @spec route_patterns_by_route_and_headsign(Stop.id_t()) :: by_route_and_headsign()
   defp route_patterns_by_route_and_headsign(stop_id) do
     stop_id
     |> RoutePatterns.Repo.by_stop_id()
-    |> Enum.reject(&not_serving_today?/1)
-    |> Enum.reject(&ends_at?(&1, stop_id))
-    |> Enum.map(&add_polyline/1)
-    |> Enum.group_by(& &1.route_id)
-    |> Enum.map(fn {k, v} -> {k, Enum.group_by(v, & &1.headsign)} end)
+    |> Stream.reject(&ends_at?(&1, stop_id))
+    |> Stream.chunk_by(& &1.route_id)
+    |> Stream.map(fn [%RoutePattern{route_id: route_id} | _] = route_route_patterns ->
+      by_headsign =
+        route_route_patterns
+        |> Stream.chunk_by(& &1.headsign)
```

Comment:
question: What is guaranteeing the sorted order?

Reply:
nothing here, this gets sorted in the frontend

Reply:
I think it would need to be sorted for `chunk_by` to work, looking at [this example](https://hexdocs.pm/elixir/1.12.3/Stream.html#chunk_by/2), the chunks are only formed for elements that are immediately next to each other. If the stream was ["Nubian", "Ashmont", "Nubian"], then the two Nubians wouldn't be grouped.

Reply:
ah shoot good point. building another good case for going back to `Enum.group_by`...